src/database_observer.py

```diff
@@ -0,0 +1,137 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The Database agent relation observer."""
+import logging
+import typing
+
+import ops
+from charms.data_platform_libs.v0.data_interfaces import (
+    DatabaseCreatedEvent,
+    DatabaseEndpointsChangedEvent,
+    DatabaseRequires,
+)
+from ops.charm import CharmBase
+from ops.framework import Object
+
+from charm_types import DatasourcePostgreSQL
+from constants import SYNAPSE_CONTAINER_NAME
+from database_client import DatabaseClient
+from exceptions import CharmDatabaseRelationNotFoundError
+
+logger = logging.getLogger(__name__)
+
+
+class DatabaseObserver(Object):
+    """The Database relation observer.
+
+    Attrs:
+        _pebble_service: instance of pebble service.
+    """
+
+    _RELATION_NAME = "database"
+
+    def __init__(self, charm: CharmBase):
+        """Initialize the observer and register event handlers.
+
+        Args:
+            charm: The parent charm to attach the observer to.
+        """
+        super().__init__(charm, "database-observer")
+        self._charm = charm
+        # SUPERUSER is required to update pg_database
+        self.database = DatabaseRequires(
+            self._charm,
+            relation_name=self._RELATION_NAME,
+            database_name=self._charm.app.name,
+            extra_user_roles="SUPERUSER",
+        )
+        self.framework.observe(self.database.on.database_created, self._on_database_created)
+        self.framework.observe(self.database.on.endpoints_changed, self._on_endpoints_changed)
+
+    @property
+    def _pebble_service(self) -> typing.Any:
+        """Return instance of pebble service.
+
+        Returns:
+            instance of pebble service or none.
+        """
+        return getattr(self._charm, "pebble_service", None)
```

Comment:
Out of curiosity, would it be correct that this cannot be injected due to circular dependencies?